// variant/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var variant_1 = require("./variant");
exports.variant = variant_1.variant;
exports.variantList = variant_1.variantList;
exports.variantFactory = variant_1.variantFactory;
exports.match = variant_1.match;
exports.partialMatch = variant_1.partialMatch;
exports.lookup = variant_1.lookup;
exports.partialLookup = variant_1.partialLookup;
exports.augment = variant_1.augment;
exports.outputTypes = variant_1.outputTypes;
exports.isOfVariant = variant_1.isOfVariant;
exports.matchLiteral = variant_1.matchLiteral;
exports.cast = variant_1.cast;
exports.narrow = variant_1.narrow;
exports.flags = variant_1.flags;
exports.keynum = variant_1.keynum;
exports.matchElse = variant_1.matchElse;
var tools_1 = require("./tools");
exports.fields = tools_1.fields;
exports.data = tools_1.data;
exports.payload = tools_1.payload;
exports.property = tools_1.property;
exports.exhaust = tools_1.exhaust;
var util_1 = require("./util");
exports.strEnum = util_1.strEnum;
const variant_2 = __importDefault(require("./variant"));
exports.default = variant_2.default;
